File: 2/2000s/26s/2685/CountTheNumberOfCompleteComponents.cpp
/*
https://leetcode.com/problems/count-the-number-of-complete-components/description/?envType=daily-question&envId=2026-07-11
*/

class Solution {
public:

    void dfs(vector<vector<int>>&gr, set<int>&seen, int &currNode, set<pair<int, int>>&curseen, vector<int>&curlst){
        curlst.push_back(currNode);
        seen.insert(currNode);
        for(auto &neighbor: gr[currNode]){
            if(seen.find(neighbor)==seen.end()){
                auto curpr1=make_pair(currNode, neighbor);
                auto curpr2=make_pair(neighbor, currNode);
                if(curseen.find(curpr1)==curseen.end()&&
                   curseen.find(curpr2)==curseen.end()){
                    curseen.insert(curpr1);
                    curseen.insert(curpr2);
                    dfs(gr, seen, neighbor, curseen, curlst);
                }
            }
        }
        return;
    }

    int countCompleteComponents(int n, vector<vector<int>>& edges) {
        int res=0;
        vector<vector<int>>gr(n);
        for(auto &edge: edges){
            gr[edge[0]].push_back(edge[1]);
            gr[edge[1]].push_back(edge[0]);
        }
        set<int>seen;
        for(int i=0; i<n; ++i){
            set<pair<int, int>>curseen;
            bool curflag=true;
            if(seen.find(i)==seen.end()){
                if(gr[i].size()==0){
                    seen.insert(i);
                } else{
                    vector<int>curlst;
                    dfs(gr, seen, i, curseen, curlst);
                    int fnNode=curlst.back();
                    vector<int>allRootsChs=gr[i];
                    allRootsChs.push_back(i);
                    sort(allRootsChs.begin(), allRootsChs.end());
                    vector<int>allFnNodesChs=gr[fnNode];
                    allFnNodesChs.push_back(fnNode);
                    sort(allFnNodesChs.begin(), allFnNodesChs.end());
                    if(allRootsChs!=allFnNodesChs){
                        curflag=false;
                    }
                    if(curflag){
                        for(auto &nei: gr[i]){
                            vector<int>nxtChsToChk=gr[nei];
                            nxtChsToChk.push_back(nei);
                            sort(nxtChsToChk.begin(), nxtChsToChk.end());
                            if(nxtChsToChk!=allRootsChs){
                                curflag=false;
                                break;
                            }
                        }
                    }
                }
                if(curflag) res++;
            }
        }
        return res;
    }
};

File: 2/2000s/26s/2685/CountTheNumberOfCompleteComponents.java
/*
https://leetcode.com/problems/count-the-number-of-complete-components/description/?envType=daily-question&envId=2026-07-11
*/

class Pair{

    public int first;

    public int second;

    Pair(int first, int second){
        this.first=first;
        this.second=second;
    }

    @Override
    public boolean equals(Object obj){
        if(this==obj) return true;
        if(obj==null || getClass()!=obj.getClass()) return false;

        Pair other=(Pair)obj;
        return first==other.first && second==other.second;
    }

    @Override
    public int hashCode(){
        return Objects.hash(first,second);
    }

}

class Solution {

    public static void depthFirstSearch_forNodes(ArrayList<ArrayList<Integer>>graph, 
                                                HashSet<Integer>seenNodes,
                                                int currentNode, 
                                                HashSet<Pair>currSeenPairNodes, 
                                                ArrayList<Integer>currentNodePath){

        currentNodePath.add(currentNode);
        seenNodes.add(currentNode);
        for(int i=0; i<graph.get(currentNode).size(); ++i){
            int currentNeighbor=graph.get(currentNode).get(i);
            if(!seenNodes.contains(currentNeighbor)){
                Pair firstPair=new Pair(currentNode, currentNeighbor);
                Pair secondPair=new Pair(currentNeighbor, currentNode);
                if(!currSeenPairNodes.contains(firstPair)&&
                   !currSeenPairNodes.contains(secondPair)){
                        currSeenPairNodes.add(firstPair);
                        currSeenPairNodes.add(secondPair);
                        depthFirstSearch_forNodes(graph, seenNodes, currentNeighbor, currSeenPairNodes, currentNodePath);
                   }
            }
        }
        return;
    }

    public int countCompleteComponents(int n, int[][] edges) {
        int finalCompConnComponents=0;
        ArrayList<ArrayList<Integer>>graph=new ArrayList<>();
        for(int i=0; i<n; ++i){
            graph.add(new ArrayList<>());
        }
        for(int edge[]: edges){
            graph.get(edge[0]).add(edge[1]);
            graph.get(edge[1]).add(edge[0]);
        }
        HashSet<Integer>seenNodes=new HashSet<>();
        for(int i=0; i<n; ++i){
            HashSet<Pair>currSeenPairNodes=new HashSet<>();
            boolean isCurrentCompConnected=true;
            if(!seenNodes.contains(i)){
                if(graph.get(i).size()==0){
                    seenNodes.add(i);
                } else{
                    ArrayList<Integer>currentNodePath=new ArrayList<>();
                    depthFirstSearch_forNodes(graph, seenNodes, i, currSeenPairNodes, currentNodePath);
                    int finalNode=currentNodePath.get(currentNodePath.size()-1);
                    ArrayList<Integer>rootWithChildrenNodes=new ArrayList<>(graph.get(i));
                    rootWithChildrenNodes.add(i);
                    Collections.sort(rootWithChildrenNodes);
                    ArrayList<Integer>finalNodesWithChildren=new ArrayList<>(graph.get(finalNode));
                    finalNodesWithChildren.add(finalNode);
                    Collections.sort(finalNodesWithChildren);
                    if(!rootWithChildrenNodes.equals(finalNodesWithChildren)){
                        isCurrentCompConnected=false;
                    }
                    if(isCurrentCompConnected){
                        for(int j=0; j<graph.get(i).size(); ++j){
                            int nxtChild=graph.get(i).get(j);
                            ArrayList<Integer> nextChildrensToCheck=new ArrayList<>(graph.get(nxtChild));
                            nextChildrensToCheck.add(nxtChild);
                            Collections.sort(nextChildrensToCheck);
                            if(!nextChildrensToCheck.equals(rootWithChildrenNodes)){
                                isCurrentCompConnected=false;
                                break;
                            }
                        }
                    }
                }
                if(isCurrentCompConnected) finalCompConnComponents++;
            }
        }
        return finalCompConnComponents;
    }
}